apis/v1alpha1/clientsettingspolicy_types.go

```diff
@@ -0,0 +1,149 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:categories=nginx-gateway-fabric,shortName=cspolicy
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+// +kubebuilder:metadata:labels="gateway.networking.k8s.io/policy=inherited"
+
+// ClientSettingsPolicy is an Inherited Attached Policy. It provides a way to configure the behavior of the connection
+// between the client and NGINX Gateway Fabric.
+type ClientSettingsPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the ClientSettingsPolicy.
+	Spec ClientSettingsPolicySpec `json:"spec"`
+
+	// Status defines the state of the ClientSettingsPolicy.
+	Status ClientSettingsPolicyStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ClientSettingsPolicyList contains a list of ClientSettingsPolicies.
+type ClientSettingsPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ClientSettingsPolicy `json:"items"`
+}
+
+// ClientSettingsPolicySpec defines the desired state of ClientSettingsPolicy.
+type ClientSettingsPolicySpec struct {
+	// TargetRef identifies an API object to apply the policy to.
+	// Object must be in the same namespace as the policy.
+	//
+	// Support: Gateway
+	TargetRef gatewayv1alpha2.PolicyTargetReference `json:"targetRef"`
+
+	// Default defines default policy configuration for the targeted resource.
+	//
+	// +optional
+	Default *ClientSettingsPolicyConfig `json:"default,omitempty"`
+}
+
+// ClientSettingsPolicyStatus defines the current state of ClientSettingsPolicy.
+type ClientSettingsPolicyStatus struct {
+	// Conditions describe the current conditions of the ClientSettingsPolicy.
+	//
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=8
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// ClientSettingsPolicyConfig contains client settings policy configuration.
+type ClientSettingsPolicyConfig struct {
+	// Body defines the client request body settings.
+	//
+	// +optional
+	Body *ClientBody `json:"body,omitempty"`
+
+	// KeepAlive defines the keep-alive settings.
+	//
+	// +optional
+	KeepAlive *ClientKeepAlive `json:"keepAlive,omitempty"`
+}
+
+// ClientBody contains the settings for the client request body.
+type ClientBody struct {
+	// MaxSize sets the maximum allowed size of the client request body.
+	// If the size in a request exceeds the configured value,
+	// the 413 (Request Entity Too Large) error is returned to the client.
+	// Setting size to 0 disables checking of client request body size.
+	// Default: 1m.
```

Comment:
would it make here and in the other CRDs to actually specify defaults so that they are set during resource creating if the field is empty?

Reply:
I intentionally did not set defaults with annotations because I think we should avoid explicitly setting the nginx directives to their default values in the config. By not setting them explicitly, if nginx changes the defaults, we will pick them up without having to make any code changes.

Reply:
> . By not setting them explicitly, if nginx changes the defaults, we will pick them up without having to make any code changes.

yep, although we will need to change the comment that tells which value is the default. so we need to keep a watch for it or something?

Reply:
> although we will need to change the comment that tells which value is the default.
I know 🙁 

Alternatively, we could link to the nginx directive documentation...


> so we need to keep a watch for it or something?
Yeah, we'd need to check if any defaults changed every release, which is a huge pain.

Reply:
> Alternatively, we could link to the nginx directive documentation...

👍